### ISSUE TITLE
[Web] empty white website in Chrome

### ISSUE THREAD

**Issue by svenefftinge** [flutter/flutter#32436]
I tried to make the Flutter for Web run in Gitpod. 
All seems to be working in Firefox and Safari (I'm on OSX) but in Chrome I don't anything in the opened website.

To reproduce start this Gitpod workspace in chrome (clicking the link):

https://gitpod.io/#https://github.com/svenefftinge/flutter_web

**Comment by yjbanov** [comment]
/cc @vsmenon it seems gitpod.io is having trouble serving DDC output. I'm not sure if it's an easy fix.

@svenefftinge I was able to run it in Chrome by running in release mode, i.e. `webdev serve -r`.

**Comment by svenefftinge** [comment]
Ah, thanks makes sense. Do you have any pointer to more information on how the chrome <-> ddc connection works? Is it common that there are issues when proxies are involved?

**Comment by grouma** [comment]
I don't think this is a DDC issue. 

This is likely an issue with `webdev`. There is an initial handshake that is done between the application and `webdev`. This handshake is required to properly support the `--debug` flag. Since you can't debug outside of Chrome we skip this handshake. 

See here:
https://github.com/dart-lang/webdev/blob/7d14c3b21a2e0e85dc9764a49050088d714f7064/webdev/web/client.dart#L93-L103

This handshake is done over an SSE connection using `package:sse`. We specifically chose SSE over  websockets because it plays better with our internal proxies.

**Comment by yjbanov** [comment]
@grouma is it possible to improve the handshake process to detect the best connection method between SSE and websockets?

**Comment by grouma** [comment]
I took a brief look at this.

The issue is in fact the result of the SSE connection. For some reason the initial handshake returns `Not Found` for the corresponding request:
https://8080-b5e38ad7-443c-4451-a6e9-17f517638931.ws-us0.gitpod.io/$sseHandler?sseClientId=03837480-5970-11e9-e61f-e19c9c88e966

I don't see webdev supporting both SSE and Websockets. SSE has strong support across the major browsers (excluding IE 11) and in my experience works better with various proxies. We should investigate this issue further and figure out why the request is not being resolved.

That being said, the SSE connection is only required for enhanced workflows like auto hot-restart etc. To unblock users we should downgrade to a more traditional experience if `--debug` isn't provided.

**Comment by pinkeshdarji** [comment]
Any update on this issue? We still have run with webdev serve --release.

**Comment by grouma** [comment]
No update. I've created a tracking bug which would be an appropriate workaround:
https://github.com/dart-lang/webdev/issues/540

It still would be beneficial understand why the requests created by the injected client are not correctly resolving in the Gitpod environment. I don't have time to investigate myself so any help there would be greatly appreciated.

**Comment by asmallkite** [comment]
Same error on the master channel. Error does not appear on the dev channel.

you can do this as fellows.

`flutter channel dev`
`flutter upgrade`
`flutter run -d chrome`

**Comment by grouma** [comment]
I fixed a race condition in `package:dwds`. @jonahwilliams you'll need to rev to the latest when [https://github.com/dart-lang/webdev/pull/682](this) is published

**Comment by DanTup** [comment]
I was just doing some testing of this (through Flutter) and the first thing I hit when trying to run was that the injected client was trying to connect to `https://localhost:8080/$sseHandler?...`. This seems to be the same thing I see in VS Online and what led to https://github.com/dart-lang/webdev/pull/822 (so I think it's required here too). I'm not sure why this didn't happen for the testing above though (I the proxies are rewriting the hostname to `localhost` since the server might not understand the public hostname that was exposed).

Changing the hostname and manually trying to hit the SSE handler in the browser gave a 404 (as Gary had above), though it seems like this is the behaviour for [Dart's SSE if the Accept header isn't `Accept: text/event-stream`](https://github.com/dart-lang/sse/blob/master/lib/server/sse_handler.dart#L117-L129). Using `curl` with the correct header seemed to connect fine.

So, I ran on a fork of dwds with my fix above and got slightly further, but because GitPod doesn't support `asExternalUri` I had to use a build of the Dart extension with the hostname hard-coded to get the right base URL into the injected client:

```ts
public async asExternalUri(uri: vs.Uri): Promise<vs.Uri> {
	if (!vs.env.asExternalUri)
		return uri.with({ authority: "8080-e91dabac-acc2-4018-9f2d-ca78c2d9f38d.ws-eu01.gitpod.io" });
}
```

With that, the initial SSE connection seems to worked:

![Screenshot 2019-11-18 at 4 12 37 pm](https://user-images.githubusercontent.com/1078012/69069840-158c4680-0a1f-11ea-9739-355c07615a4d.png)

However, it seems to drop every 60 seconds and reconnect - I'm not sure if that's expected or if there's an unwanted 60s timeout somewhere. I can't tell if that will affect general debugging as I can't easily rig that bit (I can test it more when `asExternalUri` is implemented for GitPod).

**Comment by grouma** [comment]
> However, it seems to drop every 60 seconds and reconnect - I'm not sure if that's expected or if there's an unwanted 60s timeout somewhere.

That isn't surprising. We experienced similar problems with various proxy servers internally. Many of them forcefully close longstanding connections after 60 seconds or 10 minutes. That's why we have retry logic in the client.

**Comment by DanTup** [comment]
Yeah, I suspected it might be the proxy (and on the surface, it doesn't seem crazy to drop connections that look like they've just hung open with no data). I don't know what impact that'll have on debugging though - besides performance, would all messages still make it through in both directions?

@svenefftinge does this seem like something the GitPod proxy may be doing? Is there scope to adjust it?

**Comment by grouma** [comment]
> Yeah, I suspected it might be the proxy (and on the surface, it doesn't seem crazy to drop connections that look like they've just hung open with no data). I don't know what impact that'll have on debugging though - besides performance, would all messages still make it through in both directions?

Hasn't been an issue internally. We re-establish the connection almost immediately so performance isn't a problem.

**Comment by DanTup** [comment]
Ok great, probably nothing to worry about there then :-)

**Comment by geropl** [comment]
> does this seem like something the GitPod proxy may be doing? Is there scope to adjust it?

@DanTup Yes, we do drop stale connections after 60s. We're currently looking into refining this to better support websockets and SSE.

**Comment by geropl** [comment]
I tried to reproduce the issue described here to learn what we could improve in Gitpod to better support flutter.

I basically opened Sven's link https://gitpod.io/#https://github.com/svenefftinge/flutter_web, which automatically (after a PATH issue fix) runs `webdev build && webdev serve` (flutter SDK version: `v1.9.1+hotfix.2-stable`). This worked out of the box, and showed an page with `Hello World` (tested Firefox and Chrome).

@DanTup I neither hit the `localhost` issue nor the `asExternalUri` issue (although I admit the later would be really useful). Could you elaborate on your setup & command used?

Does that mean the issue Sven originally described has been fixed in the meantime? Or does the SSE connection provide further functionality that I'm not aware of and which is still missing? I tried it locally, too, but it looked identical to the result within Gitpod.

**Comment by DanTup** [comment]
@geropl yep, the stuff that uses SSE will provide additional functionality. Testing is tricky right now as it needs forks of Flutter & webdev with some unmerged changes (and a custom build of the Dart extension).

There are two significant blocking issues I know of:

- `asExternalUri` support on GitPod (I think the change was merged into theia but I don't know what's involved to get it to GitPod). This is needed so that we can automatically expose ports and also get at the exposed URL (we need to inject some into javascript, so they need to be the public URLs and not internal)
- proxy is dropping port numbers from the `HOST` header and also providing `x-forwarded-proto` but not `x-forwarded-host` (raised at https://github.com/gitpod-io/gitpod/issues/950). The result is that when Dart tries to reconstruct [`requestedUri`](https://api.flutter.dev/flutter/dart-io/HttpRequest/requestedUri.html) it comes up with an invalid URL (for ex. https://localhost/).

Once those are resolved, I should be able to build a repository that allows for proper testing (though still probably using some custom forks of Flutter/etc. until those changes are all landed).

**Comment by geropl** [comment]
@DanTup Thanks for the clarification. I will push proxy fixes and have a look at https://github.com/gitpod-io/gitpod/issues/950 then, I think the `asExternalUri` already hit master.

**Comment by ferhatb** [comment]
@geropl can we close issue?

**Comment by DanTup** [comment]
There are a few outstanding issues I know of for debugging to work on GitPod:

- Enable debug extension for web-server device (https://github.com/flutter/flutter/pull/53765)
- Add a flag to Flutter to allow forcing WebSockets instead of SSE for the debug service proxy
- Pass flags through from VS Code extension

Top one is in progress, and the other two I'm hoping to look at soon once that lands.
